pandas/core/index.py

```diff
@@ -3260,8 +3260,12 @@ def _evaluate_compare(self, other):
                 elif isinstance(other, Index):
                     other = self._create_categorical(self, other.values, categories=self.categories, ordered=self.ordered)
 
+                if isinstance(other, (ABCCategorical, np.ndarray, ABCSeries)):
+                    if len(self.values) != len(other):
+                        raise ValueError("Lengths must match to compare")
+
                 if isinstance(other, ABCCategorical):
-                    if not (self.values.is_dtype_equal(other) and len(self.values) == len(other)):
```

Comment:
right ``ValueError`` is fine (thought he ``is_dtype_equal`` will raise a ``TypeError`` for invalid dtypes, which is correct as well)